[TensorPipe] Improve serialization

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39012 [RPC] Fix flaky test by waiting for async rref calls
* #39011 [TensorPipe] Detect duplicate worker names
* **#39010 [TensorPipe] Improve serialization**
* #38934 [TensorPipe] Close and join TP context at shutdown
* #38933 [TensorPipe] Implement join correctly
* #38930 [TensorPipe] Always complete futures from thread pool
* #38931 [TensorPipe] Do not mark future messages as complete after they have timed out
* #38929 [TensorPipe] Fix lock inversion upon response read error handling
* #38928 [TensorPipe] Fix timeout computation
* #38927 [TensorPipe] Add cases for TP in RPC test helpers
* #38926 [TensorPipe] Pass names of endpoints to context/pipe for easier debugging

The initial version of the serialization for the TensorPipe RPC agent (i.e., the conversion from rpc::Message to tensorpipe::Message) worker around a limitation of TensorPipe of only allowing one payload per message by pickling each tensor separately and storing the pickles as metadata (which is a less efficient way of sending data over, as it goes through more copies). Having now lifter that limitation we can now improve the way we serialize. We now put the type and the id as their own payloads, we do a single pickling pass for all the tensors of the message (which allows us to deduplicate them) and store the pickle as a payload. My impression is that pickling is a somewhat costly operation, so reducing the number of times we do it should be beneficial for performance. For this same reason, another change I've done here is separate the allocation of the buffers from the deserialization. This will allow us (in the future) to perform the allocation on the I/O event loop but perform the unpickling in the worker thread, thus keeping the event loop more responsive.

Differential Revision: [D21716067](https://our.internmc.facebook.com/intern/diff/D21716067/)

lw committed:
[TensorPipe] Improve serialization

The initial version of the serialization for the TensorPipe RPC agent (i.e., the conversion from rpc::Message to tensorpipe::Message) worker around a limitation of TensorPipe of only allowing one payload per message by pickling each tensor separately and storing the pickles as metadata (which is a less efficient way of sending data over, as it goes through more copies). Having now lifter that limitation we can now improve the way we serialize. We now put the type and the id as their own payloads, we do a single pickling pass for all the tensors of the message (which allows us to deduplicate them) and store the pickle as a payload. My impression is that pickling is a somewhat costly operation, so reducing the number of times we do it should be beneficial for performance. For this same reason, another change I've done here is separate the allocation of the buffers from the deserialization. This will allow us (in the future) to perform the allocation on the I/O event loop but perform the unpickling in the worker thread, thus keeping the event loop more responsive.

Differential Revision: [D21716067](https://our.internmc.facebook.com/intern/diff/D21716067/)

[ghstack-poisoned]

lw committed:
Update on "[TensorPipe] Improve serialization"


The initial version of the serialization for the TensorPipe RPC agent (i.e., the conversion from rpc::Message to tensorpipe::Message) worker around a limitation of TensorPipe of only allowing one payload per message by pickling each tensor separately and storing the pickles as metadata (which is a less efficient way of sending data over, as it goes through more copies). Having now lifter that limitation we can now improve the way we serialize. We now put the type and the id as their own payloads, we do a single pickling pass for all the tensors of the message (which allows us to deduplicate them) and store the pickle as a payload. My impression is that pickling is a somewhat costly operation, so reducing the number of times we do it should be beneficial for performance. For this same reason, another change I've done here is separate the allocation of the buffers from the deserialization. This will allow us (in the future) to perform the allocation on the I/O event loop but perform the unpickling in the worker thread, thus keeping the event loop more responsive.

Differential Revision: [D21716067](https://our.internmc.facebook.com/intern/diff/D21716067/)

[ghstack-poisoned]

mrshenli reviewed (changes requested):
Lint failure is real:

```
  {
    path: 'torch/csrc/distributed/rpc/utils.cpp',
    start_line: 372,
    end_line: 372,
    start_column: 7,
    end_column: 7,
    annotation_level: 'failure',
    message: '[cppcoreguidelines-pro-type-const-cast] warning: do not use const_cast'
  },
  {
    path: 'torch/csrc/distributed/rpc/utils.cpp',
    start_line: 375,
    end_line: 375,
    start_column: 40,
    end_column: 40,
    annotation_level: 'failure',
    message: "[bugprone-use-after-move] warning: 'rpcMessage' used after it was moved"
  },
  {
    path: 'torch/csrc/distributed/rpc/utils.cpp',
    start_line: 399,
    end_line: 399,
    start_column: 11,
    end_column: 11,
    annotation_level: 'failure',
    message: '[cppcoreguidelines-pro-type-const-cast] warning: do not use const_cast'
  }
```

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -176,20 +176,31 @@ void TensorPipeAgent::pipeRead(
   pipe->readDescriptor([fn{std::move(fn)}, pipe](
                            const tensorpipe::Error& error,
-                           tensorpipe::Message&& tpMessage) mutable {
+                           tensorpipe::Message tpMessage) mutable {
```
Any reason for dropping the rvalue reference type requirement here?

lw replied:
The type expected by the `readDescriptor` function takes the Message by value rather than by rvalue-ref, so I found it weird to use a different type here. The two are compatible, so it makes no real difference. In fact, the Message type is movable-only, so...

If the style guide in PyTorch is to specify the rvalue ref in arguments that are expected to be moved in I can change this back.

mrshenli replied:
I see. All good, no need to change.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -176,20 +176,31 @@ void TensorPipeAgent::pipeRead(
             const tensorpipe::Error& error,
-            tensorpipe::Message&& /* unused */) mutable {
+            tensorpipe::Message tpMessage) mutable {
+          if (error) {
```
IIUC, this is the shortcut to avoid deserialization when there is an error?

lw replied:
Yes, but this was actually almost unintentional. The real benefit will be had when the deserialization will be deferred to the thread pool.

mrshenli commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -353,119 +353,137 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
-  memcpy(&mId, tpMessage.metadata.data() + sizeof(int64_t), sizeof(int64_t));
-  MessageType mType = static_cast<MessageType>(mtypei);
+  buffers.type = std::make_unique<MessageType>();
+  tpMessage.payloads[0].data = buffers.type.get();
```
Could you please elaborate more on this change? It looks like we no longer expect `tpMessage` to contain metadata?

lw replied:
This unfortunately stems from a bad naming in TensorPipe: what is called "metadata" in the tp::Message should really have been called something like "allocation hints", i.e., it should only contain information needed in order to correctly allocate the receiver's buffers. It should not be used to store what the agent considers to be a rpc::Message's metadata, but instead it should store the metadata that corresponds to tp::Message's data. The rpc::Message type and id are just some additional payload from TensorPipe's perspective, and should be stored as such: payloads are transferred more efficiently than metadata (even though here we're just talking about a few bytes...).

The moment when we'll really need to use the "allocation hints" is for GPU tensors, as in that case the allocation hints will contain the device, which we need to know when allocating buffers

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -199,17 +210,17 @@ void TensorPipeAgent::pipeWrite(
-      [rpcMessage{std::move(rpcMessage)},
-       reservedTensors{std::move(tpEntry.reservedTensors)},
-       copiedTensors{std::move(tpEntry.copiedTensors)},
+      [tpBuffers{
```
Looks like the only reason we capture the `tpBuffers` here is to keep those buffers alive? If so, why not letting `tpMessage` take the ownership?

lw replied:
That's true, we just capture tpBuffers to keep the memory alive. The tp::Message offers no way to take ownership of data. We used to have something like that but it turned out it's rather complicated: the original data could be managed in various forms (the ones we have here are `std::vector<char>`, `std::string` and `std::unique_ptr<...>`, but there exist so many more) and TP cannot support all of them. One way this could work is through `std::unique_ptr` as it supports a custom destructor,  so all the above ways could be "converted" to that. However that destructor must be specified as a template parameter, so in order to be generic this would have to be something like `std::unique_ptr<uint8_t, std::function<void(uint8_t*)>>`. And, at this point, I'm not sure it would really be more convenient for the agent to create a custom destructor rather than capture it in the lambda it's already creating anyways.

mrshenli replied:
I see.

mrshenli commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -353,119 +353,137 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
-  return TensorPipeEntry{std::move(tpMessage),
-                         std::move(reservedTensors),
-                         std::move(copiedTensors)};
+  return std::make_tuple(std::move(tpMessage), std::move(buffers));
```
What is the reason for explicitly returning `buffers`? Is the returned `buffers` used anywhere besides making sure the lifetime of payload (etc.) is correct?

lw replied:
The buffers need to be returned so that the agent can keep it alive until the write has completed. There is no other purpose. See comment above for more discussion.

mrshenli commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -353,119 +353,137 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+  tpMessage.payloads[2].data = buffers.payload.data();
+
+  buffers.pickle.resize(tpMessage.payloads[3].length);
+  tpMessage.payloads[3].data = buffers.pickle.data();
```
Shall we have some `constexpr` for these indices instead of hard coding them?

lw replied:
Good call!

lw commented:
TBH, I disagree with the linter errors:
- one of the `const_cast`s is bad but necessary: the rpc::Message stores the payload as std::vector<char> so if we want to avoid a copy we need to first allocate a std::vector of the right size and then _read into it_. For that we need to get a non-const pointer to its data but the stdlib only allows to get a const pointer, so we need const_cast. A better solution would be to use a different type to store the payload on the rpc::Message, something like `std::unique_ptr<uint8_t[]>` (which however means that the length has to be stored separately).
- the other `const_cast` could be avoided but with some more changes: at the moment TensorPipe uses the same Message class for reading and writing (perhaps we should change that, but it would lead to quite a bit of duplication...) so its pointers are non-const. Therefore when getting the pointer from the rpc::Message::payload() we need to cast it to non-const.
- the "use after move" is something I don't like but which I couldn't find a way around. The only way to _move_ the payload out of rpc::Message is to use `movePayload`, which however is `&&`-qualified, so it only works if we move the message, which leads to the weird expression `std::move(rpcMessage).movePayload()`. However, after we move the payload out, we still want to access `rpcMessage.tensors()`, which leads to the use-after-move. We could fix it by doing these operations in another order, but I think a more robust fix here would be to change the API of rpc::Message, for example by removing the `&&` qualification from rpc::Message.

lw committed:
Update on "[TensorPipe] Improve serialization"


The initial version of the serialization for the TensorPipe RPC agent (i.e., the conversion from rpc::Message to tensorpipe::Message) worker around a limitation of TensorPipe of only allowing one payload per message by pickling each tensor separately and storing the pickles as metadata (which is a less efficient way of sending data over, as it goes through more copies). Having now lifter that limitation we can now improve the way we serialize. We now put the type and the id as their own payloads, we do a single pickling pass for all the tensors of the message (which allows us to deduplicate them) and store the pickle as a payload. My impression is that pickling is a somewhat costly operation, so reducing the number of times we do it should be beneficial for performance. For this same reason, another change I've done here is separate the allocation of the buffers from the deserialization. This will allow us (in the future) to perform the allocation on the I/O event loop but perform the unpickling in the worker thread, thus keeping the event loop more responsive.

Differential Revision: [D21716067](https://our.internmc.facebook.com/intern/diff/D21716067/)

[ghstack-poisoned]

mrshenli commented:
> TBH, I disagree with the linter errors:

For the const-cast, we can add `// NOLINT(cppcoreguidelines-pro-type-const-cast)` to the corresponding lines to shut the linter. But we cannot land this as-is to break master.

mrshenli commented:
> We could fix it by doing these operations in another order, 

Reordering sounds good to me to fix lint in this PR. 

> but I think a more robust fix here would be to change the API of rpc::Message, for example by removing the && qualification from rpc::Message.

This means we need some additional field in `rpc::Message` to indicate whether the `payload` and `tensors` have been moved?

mrshenli commented:
Will drop "request changes" when linters are happy.

lw commented:
About moving the rpc::Message, I took another look at that class and it seems that for accessing the vector of tensors there are two methods: a const-qualified one that returns a const reference, and a unqualified one that returns a non-const reference. For the payload, however, there is only the const-qualified one. I thus think that adding a unqualified one makes sense and it fixes the issue at hand (since we can move out from a non-const reference).

As a general design consideration, if we consider the move operation to be "destructive" (it prevents from using the moved-out object later on), then the rpc::Message looks a bit weird to me because it is _one_ object that owns _two_ other object, and in order to move either one of the sub-objects out one needs to destroy the Message, thus preventing from moving out the other sub-object. One solution is the the one above using non-const refs, where we allow to move out the sub-objects without destroying the Message. Another solution would be to move both sub-objects out in a single call, like a `std::tuple<std::vector<char>, std::vector<at::Tensor>> movePayloadAndTensors()` method.

lw committed:
Update on "[TensorPipe] Improve serialization"


The initial version of the serialization for the TensorPipe RPC agent (i.e., the conversion from rpc::Message to tensorpipe::Message) worker around a limitation of TensorPipe of only allowing one payload per message by pickling each tensor separately and storing the pickles as metadata (which is a less efficient way of sending data over, as it goes through more copies). Having now lifter that limitation we can now improve the way we serialize. We now put the type and the id as their own payloads, we do a single pickling pass for all the tensors of the message (which allows us to deduplicate them) and store the pickle as a payload. My impression is that pickling is a somewhat costly operation, so reducing the number of times we do it should be beneficial for performance. For this same reason, another change I've done here is separate the allocation of the buffers from the deserialization. This will allow us (in the future) to perform the allocation on the I/O event loop but perform the unpickling in the worker thread, thus keeping the event loop more responsive.

Differential Revision: [D21716067](https://our.internmc.facebook.com/intern/diff/D21716067/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM!

For the moving Message field, I'd vote for the second option:

> Another solution would be to move both sub-objects out in a single call, like a std::tuple<std::vector<char>, std::vector<at::Tensor>> movePayloadAndTensors() method.
